katdal/applycal.py

```diff
@@ -216,57 +219,88 @@ def calc_gain_correction(sensor, index):
+    per product type and per input in the SensorCache `cache`, with template
+    'Calibration/Corrections/`cal_stream`/{product_type}/{inp}'. The virtual
+    sensor function picks the appropriate correction calculator based on the
+    cal product type, which also uses auxiliary info like the channel
+    frequencies, `data_freqs`.
 
-    This maps receptor inputs to the relevant indices in each calibration
-    product based on the ants and pols found in `attrs`. It then registers
-    a virtual sensor per input and per cal product in the SensorCache `cache`,
-    with template 'Calibration/{inp}_correction_{product}'. The virtual sensor
-    function picks the appropriate correction calculator based on the cal
-    product name, which also uses auxiliary info like the channel frequencies,
-    `data_freqs`.
+    Parameters
+    ----------
+    cache : :class:`~katdal.sensordata.SensorCache` object
+        Sensor cache serving cal product sensors and receiving correction sensors
+    attrs : dict-like
+        Calibration stream attributes (e.g. a "cal" telstate view)
+    data_freqs : array of float, shape (*F*,)
+        Centre frequency of each frequency channel of visibilities, in Hz
+    cal_stream : string, optional
+        Name of (possibly virtual) calibration stream (e.g. "L1")
+    cal_substreams : sequence of string, optional
+        Names of actual underlying calibration streams (e.g. ["cal"]),
+        defaults to [`cal_stream`] itself
     """
-    cal_ants = attrs.get('cal_antlist', [])
-    cal_pols = attrs.get('cal_pol_ordering', [])
+    if cal_substreams is None:
+        cal_substreams = [cal_stream]
+    cal_ants = attrs.get('antlist', [])
+    cal_pols = attrs.get('pol_ordering', [])
     cal_input_map = {ant + pol: (pol_idx, ant_idx)
                      for (pol_idx, pol) in enumerate(cal_pols)
                      for (ant_idx, ant) in enumerate(cal_ants)}
     if not cal_input_map:
         return
     try:
-        cal_spw = SpectralWindow(attrs['cal_center_freq'], None,
-                                 attrs['cal_n_chans'], sideband=1,
-                                 bandwidth=attrs['cal_bandwidth'])
+        cal_spw = SpectralWindow(attrs['center_freq'], None,
+                                 attrs['n_chans'], sideband=1,
+                                 bandwidth=attrs['bandwidth'])
         cal_freqs = cal_spw.channel_freqs
     except KeyError:
-        logger.warning('Missing cal spectral attributes, disabling applycal')
+        logger.warning("Disabling cal stream '%s' due to missing "
+                       "spectral attributes", cal_stream)
         return
 
-    def calc_correction_per_input(cache, name, inp, product):
+    def indirect_cal_product(cache, name, product_type):
+        # XXX The first underscore below is actually a telstate separator...
+        return cache.get('{}_product_{}'.format(cal_substreams[0], product_type))
```

Comment:
If you want to avoid the separator, could you instead use `attrs.get('product_{}'.format(product_type))`? That would bypass the cache of the raw attribute, but it presumably still be cached in the virtual sensor? And for now it would also remove the need to pass cal_substreams at all, although I assume you have other plans for it for L2.

Or are you trying to avoid conflating the use of `attrs` for metadata with access to the solutions?

Reply:
The problem is that `attrs['product_G']` is a telstate mutable on which you need to do `.get_range`, while `cache['cal_product_G']` is a `SensorData` wrapper that can be extracted to a `CategoricalData`, which is what you really need in the applycal machinery. And `cache` does not have views (yet), which leads to this code. I wanted to avoid sensor extraction inside `applycal`.  (And yes, L2 will need the substreams.)